docs/content/docs/dev/table/data_stream_api.md

```diff
@@ -0,0 +1,2106 @@
+            User.class,
+            DataTypes.FIELD("name", DataTypes.STRING()),
+            DataTypes.FIELD("score", DataTypes.INT()),
+            DataTypes.FIELD("event_time", DataTypes.TIMESTAMP_LTZ(3))));
+```
+{{< /tab >}}
+{{< tab "Scala" >}}
+```scala
+import org.apache.flink.streaming.api.scala.DataStream
+import org.apache.flink.table.api.DataTypes
+
+case class User(name: String, score: java.lang.Integer, event_time: java.time.Instant)
+
+tableEnv.executeSql(
+    "CREATE TABLE GeneratedTable "
+    + "("
+    + "  name STRING,"
+    + "  score INT,"
+    + "  event_time TIMESTAMP_LTZ(3),"
+    + "  WATERMARK FOR event_time AS event_time - INTERVAL '10' SECOND"
+    + ")"
+    + "WITH ('connector'='datagen')")
+
+val table = tableEnv.from("GeneratedTable")
+
+
+// === EXAMPLE 1 ===
+
+// use the default conversion to instances of Row
+
+// since `event_time` is a single rowtime attribute, it is inserted into the DataStream
+// metadata and watermarks are propagated
+
+val dataStream: DataStream[Row] = tableEnv.toDataStream(table)
+
+
+// === EXAMPLE 2 ===
+
+// a data type is extracted from class `User`,
+// the planner reorders fields and inserts implicit casts where possible to convert internal
+// data structures to the desired structured type
+
+// since `event_time` is a single rowtime attribute, it is inserted into the DataStream
+// metadata and watermarks are propagated
+
+val dataStream: DataStream[User] = tableEnv.toDataStream(table, User.class)
+
+// data types can be extracted reflectively as above or explicitly defined
+
+val dataStream: DataStream[User] =
+    tableEnv.toDataStream(
+        table,
+        DataTypes.STRUCTURED(
+            User.class,
+            DataTypes.FIELD("name", DataTypes.STRING()),
+            DataTypes.FIELD("score", DataTypes.INT()),
+            DataTypes.FIELD("event_time", DataTypes.TIMESTAMP_LTZ(3))))
+```
+{{< /tab >}}
+{{< /tabs >}}
+
+Note that only non-updating tables are supported by `toDataStream`. Usually, time-based operations
+such as windows, interval joins, or the `MATCH_RECOGNIZE` clause are a good fit for insert-only
+pipelines next to simple operations like projections and filters. Pipelines with operations that
+produce updates can use `toChangelogStream`.
+
+{{< top >}}
+
+Handling of Changelog Streams
+-----------------------------
+
+Internally, Flink's table runtime is a changelog processor. The concepts page describes how
+[dynamic tables and streams are related]({{< ref "docs/dev/table/concepts/dynamic_tables" >}})
+with each other.
+
+A `StreamTableEnvironment` offers the following methods to expose these _change data capture_ (CDC) functionalities:
+
+- `fromChangelogStream(DataStream)`: Interprets a stream of changelog entries as a table. The stream
+record type must be `org.apache.flink.types.Row` as the `RowKind` flag that is contained in
+every record is evaluated during runtime. Event-time and watermarks are not propagated by default.
```

Comment:
```suggestion
record type must be `org.apache.flink.types.Row` since its `RowKind` flag is evaluated at runtime. Event-time and watermarks are not propagated by default.
```